### PR TITLE
Don’t warn for types used in trait implementation

### LINKED ISSUE #605: linkedlist lint should not warn when implementing a trait for LinkedList
Running clippy on serde results in the following warning:

```
src/ser/impls.rs:291:23: 291:36 warning: I see you're using a LinkedList! Perhaps you meant some other data structure?, #[warn(linkedlist)] on by default
src/ser/impls.rs:291 impl<T> Serialize for LinkedList<T>
                                           ^~~~~~~~~~~~~
src/ser/impls.rs:291:23: 291:36 help: a VecDeque might work
for further information visit https://github.com/Manishearth/rust-clippy/wiki#linkedlist
```

Ignoring my general reservations against this lint, a trait impl on `LinkedList` isn't the same as using a linked list, so I think we should ignore cases like this.

### PR DESCRIPTION
**Pull request by mcarton**
Fix #605. Ref #553.

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -47,27 +48,63 @@ impl LintPass for TypePass {
 }
 
 impl LateLintPass for TypePass {
-    fn check_ty(&mut self, cx: &LateContext, ast_ty: &Ty) {
```
Comment (Manishearth) [inline src/types.rs]: Can we instead check the parent here using the node map?

Comment (mcarton) [inline src/types.rs]: I tried that but while `cx.tcx.map.get_parent(ast_ty.id)` does not return `DUMMY_NODE_ID`, `cx.tcx.map.find(cx.tcx.map.get_parent(ast_ty.id))` is always `None` here.

Comment (Manishearth) [inline src/types.rs]: You want `get_parent_node` here

Comment (mcarton) [inline src/types.rs]: Nope, it returns the same ID here.

**Comment by llogiq** [comment]
@mcarton @Manishearth How's the status on this PR? Is something needed or should we merge?

**Comment by llogiq** [comment]
@mcarton @Manishearth ping?

**Comment by Manishearth** [comment]
Needs a rebase, but otherwise 👍

**Comment by oli-obk** [comment]
closed in favour of #1393

**Pull request closed without merging**